pkg/util/validation/errors.go

```diff
@@ -0,0 +1,191 @@
+package validation
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+)
+
+// ValidationError is an error returned by series validation. It provides a function to convert
+// the error into an HTTPGRPC error.
+//
+// nolint:golint ignore stutter warning
+type ValidationError interface {
+	error
+
+	// ToHTTPGRPCError returns the httpgrpc version of the error.
+	ToHTTPGRPCError() error
```

Comment:
Nit: It would be nicer if `ValidationError` interface was not defined in terms of specific transport mechanism. Instead it could simply have a method `IsValidationError() bool` [or some other that makes sense]. Call to `httpgrpc.Errorf` could then be done in distributor, by using message from `err.Error()` function, which would do the formatting. Duplicity in that all validation errors return same `StatusBadRequest` code would be removed.

Reply:
Sounds a good idea. Done!